p2p-specs/p2p-interface.md

```diff
@@ -0,0 +1,229 @@
+| `NewPooledUserOperationsHashes`  | ``                        |
+
+
+Bundlers MUST reject (fail validation) messages containing an incorrect type, or invalid payload.
+
+When processing incoming gossip, Bundlers MAY de-score or disconnect peers who fail to observe these constraints.
+
+For any optional queueing, Bundlers SHOULD maintain maximum queue sizes to avoid DoS vectors.
+
+
+#### Global topics
+
+There are two primary global topics used to propagate user operations (`UserOperationWithEntryPoint`)
+and aggregate user operation hashes (`NewPooledUserOperationsHashes`) to all nodes on the network.
+
+##### `UserOperationWithEntryPoint`
+
+The `UserOperationWithEntryPoint` topic is the concatenation of EntryPoint address and UserOperation message serialized using SSZ
+
+##### `NewPooledUserOperationsHashes`
+
+The `NewPooledUserOperationsHashes` topic is used solely for propagating to all the connected nodes on the networks. One or more UserOps that have appeared in the network and which have not yet been included in a block are propagated to a fraction of the nodes connected to the network.
+
+
+##### `GetPooledUserOps`
+
+The `GetPooledUserOps` requests UserOps from the recipients UserOp mempool for a given EntryPoint contract address. The recommended soft limit for GetPooledUserOps requests is 256 hashes (8 KiB). The recipient may enforce an arbitrary limit on the response (size or serving time), which must not be considered a protocol violation.
+
+##### `PooledUserOperations`
+
+The `PooledUserOperations` is a response to the `GetPooledUserOps`, returning the requested UserOperations from the local pool. The items in the list are UserOps in the format described in ERC4337 specification. 
+
+## Constants
+
+| Name                             | Message Type              |
+|----------------------------------|---------------------------|
+| `MAX_OPS_PER_REQUEST`            | `uint64(2**10)` = 1024    |
+
+
+## Container Specifications
+
+The following types are SimpleSerialize (SSZ) containers.
+
+#### `UserOp`
+
+```python
+class UserOp(Container):
+    sender: Address
+    nonce: uint256
+    init_code: bytes
+    call_data: bytes
+    call_gas_limit: uint256
+    verification_gas_limit: uint256
+    pre_verification_gas: uint256
+    max_fee_per_gas: uint256
+    max_priority_fee_per_gas: uint256
+    paymaster_and_data: bytes
+    signature: bytes
+```
+
+#### `UserOperationWithEntryPoint`
+
+```python
+class UserOperationWithEntryPoint(Container):
+    entry_point_contract: Address
+    verified_at_block_hash: uint256
+    chain_id: uint256
+    user_operations: List[UserOp, MAX_OPS_PER_REQUEST]
+```
+
+#### `GetPooledUserOps`
+
+```python
+class GetPooledUserOps(Container):
+    request_id: uint256
+    chain_id: uint256
+    entry_point_address: List[Address, MAX_OPS_PER_REQUEST]
+```
+
+#### `PooledUserOperations`
+
+```python
+class PooledUserOperations(Container):
+    chain_id: uint256
+    entry_point_address: Address
```

Comment:
this is a list of UserOperationWithEntryPoint, not of UserOp.
(and thus not needing entry_point as a separate field.

Reply:
Typically there would only be one entrypoint.  Doesn't it make sense to ask for a specific entrypoint in the request (possibly sending multiple requests if multiple entrypoints are supported by the requester), and return a PooledUserOperations with only one entrypoint (listed once) and UserOps rather than UserOperationsWithEntryPoint?